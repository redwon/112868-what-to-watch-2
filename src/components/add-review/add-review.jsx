import React, {Fragment} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';

import {MovieType} from '../../types';
import {Operations} from '../../reducer/reviews/reviews';
import {getMovieById} from '../../selectors';

import SignIn from '../sign-in/sign-in';

const AddReview = (props) => {
  if (props.isAuthorizationRequired) {
    return <SignIn />;
  }

  if (!props.movie) {
    return null;
  }

  const {
    movie,
    comment,
    isCommentValid,
    commentErrorMessage,
    rating,
    isRatingValid,
    ratingErrorMessage,
    isFormValid,
    isShowError,
    onShowError,
    onUserInput,
    onAddReview
  } = props;

  const ratings = [1, 2, 3, 4, 5];

  const onSubmitHandler = (evt) => {
    evt.preventDefault();
    onShowError(true);

    if (!isFormValid) {
      return;
    }

    onAddReview(rating, comment, movie.id);
  };

  return (
    <section className="movie-card movie-card--full">
      <div className="movie-card__header">
        <div className="movie-card__bg">
          <img src={movie.backgroundImage} alt={movie.name} />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <header className="page-header">
          <div className="logo">
            <a href="main.html" className="logo__link">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </a>
          </div>

          <nav className="breadcrumbs">
            <ul className="breadcrumbs__list">
              <li className="breadcrumbs__item">
                <a href="movie-page.html" className="breadcrumbs__link">
                  {movie.name}
                </a>
              </li>
              <li className="breadcrumbs__item">
                <a className="breadcrumbs__link">Add review</a>
              </li>
            </ul>
          </nav>

          <div className="user-block">
            <div className="user-block__avatar">
              <img src="img/avatar.jpg" alt="User avatar" width="63" height="63" />
            </div>
          </div>
        </header>

        <div className="movie-card__poster movie-card__poster--small">
          <img
            src={movie.posterImage}
            alt={movie.name}
            width="218"
            height="327"
          />
        </div>
      </div>

      <div className="add-review">
        <form className="add-review__form" onSubmit={onSubmitHandler}>
          <div className="rating">
            <div className="rating__stars">
              {ratings.map((it, i) => (
                <Fragment key={i}>
                  <input
                    className="rating__input"
                    id={`star-${it}`}
                    type="radio"
                    name="rating"
                    value={it}
                    checked={it === rating}
                    onChange={onUserInput}
                  />
                  <label
                    className="rating__label"
                    htmlFor={`star-${it}`}
                    style={{opacity: (it <= rating) ? 1 : 0.5}}
                  >
                    Rating {it}
                  </label>
                </Fragment>
              ))}
            </div>
          </div>

          <div className="add-review__text">
            <textarea
              className="add-review__textarea"
              name="comment"
              placeholder="Review text"
              value={comment}
              onChange={onUserInput}
            ></textarea>

            <div className="add-review__submit">
              <button
                className="add-review__btn"
                type="submit"
              >
                Post
              </button>
            </div>
          </div>

          {isShowError && !isRatingValid && (
            <strong>{ratingErrorMessage}<br /></strong>
          )}

          {isShowError && !isCommentValid && (
            <strong>{commentErrorMessage}</strong>
          )}

        </form>
      </div>
    </section>
  );
};

AddReview.propTypes = {
  movie: MovieType,
  comment: PropTypes.string,
  isCommentValid: PropTypes.bool,
  commentErrorMessage: PropTypes.string,
  rating: PropTypes.number,
  isRatingValid: PropTypes.bool,
  ratingErrorMessage: PropTypes.string,
  isFormValid: PropTypes.bool,
  isShowError: PropTypes.bool,
  onShowError: PropTypes.func,
  onUserInput: PropTypes.func,
  onAddReview: PropTypes.func,
  isAuthorizationRequired: PropTypes.bool,
  match: PropTypes.object,
};

const mapStateToProps = (state, props) => Object.assign({}, props, {
  movie: getMovieById(state, props.match.params.id),
  isAuthorizationRequired: state.isAuthorizationRequired
});

const mapDispatchToProps = (dispatch) => ({
  onAddReview: (rating, comment, movieId) => {
    dispatch(Operations.addReview(rating, comment, movieId));
  }
});

export {AddReview};

export default connect(mapStateToProps, mapDispatchToProps)(AddReview);
