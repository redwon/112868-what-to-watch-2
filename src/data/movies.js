const movies = [
  {
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    img: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`
  },
  {
    name: `Bohemian Rhapsody`,
    img: `img/bohemian-rhapsody.jpg`
  },
  {
    name: `Macbeth`,
    img: `img/macbeth.jpg`
  },
  {
    name: `Mindhunter`,
    img: `img/mindhunter.jpg`
  }
];

export default movies;
