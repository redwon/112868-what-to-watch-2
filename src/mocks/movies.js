const movies = [
  {
    id: 1,
    name: `War of the Worlds`,
    posterImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg`,
    previewImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg`,
    backgroundImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg`,
    backgroundColor: `#9B7E61`,
    description: `Lorem ipsum`,
    rating: 9.3,
    scoresCount: 386834,
    director: `Steven Spielberg`,
    starring: [`Tom Cruise`, `Dakota Fanning`, `Tim Robbins`],
    runTime: 116,
    genre: `Adventure`,
    released: 2005,
    isFavorite: false,
    videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
  },
  {
    id: 2,
    name: `Bronson`,
    posterImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/bronson.jpg`,
    previewImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/bronson.jpg`,
    backgroundImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/bronson.jpg`,
    backgroundColor: `#73B39A`,
    description: `Lorem ipsum`,
    rating: 3.6,
    scoresCount: 109661,
    director: `Nicolas Winding Refn`,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    runTime: 92,
    genre: `Action`,
    released: 2008,
    isFavorite: false,
    videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
  }
];

export default movies;
