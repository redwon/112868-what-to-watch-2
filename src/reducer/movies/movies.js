export const initialState = [];

export const ActionType = {
  LOAD_MOVIES: `LOAD_MOVIES`,
};

export const ActionCreator = {
  loadMovies: (movies) => ({
    type: ActionType.LOAD_MOVIES,
    payload: movies
  }),
};

export const Operations = {
  loadMovies: () => (dispatch, _, api) => {
    return api.get(`/films`)
      .then((response) => {
        dispatch(ActionCreator.loadMovies(response.data));
      });
  }
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_MOVIES:
      return action.payload;
  }

  return state;
};
