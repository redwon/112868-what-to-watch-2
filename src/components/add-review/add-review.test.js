import React from 'react';
import renderer from 'react-test-renderer';

import movies from '../../mocks/movies';

import {AddReview} from './add-review';

it(`renders correctly`, () => {
  const tree = renderer
    .create(<AddReview
      movie={movies[0]}
      comment={`comment`}
      isCommentValid={true}
      rating={1}
      isRatingValid={true}
      isFormValid={true}
      isShowError={true}
      onShowError={jest.fn()}
      onUserInput={jest.fn()}
      onAddReview={jest.fn()}
      isAuthorizationRequired={false}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
