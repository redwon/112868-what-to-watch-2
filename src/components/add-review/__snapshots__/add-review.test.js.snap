// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="War of the Worlds"
        src="https://htmlacademy-react-2.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg"
      />
    </div>
    Header
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="War of the Worlds"
        height="327"
        src="https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg"
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      className="add-review__form"
      onSubmit={[Function]}
    >
      <div
        className="rating"
      >
        <div
          className="rating__stars"
        >
          <input
            checked={true}
            className="rating__input"
            id="star-1"
            name="rating"
            onChange={[MockFunction]}
            type="radio"
            value={1}
          />
          <label
            className="rating__label"
            htmlFor="star-1"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            Rating 
            1
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-2"
            name="rating"
            onChange={[MockFunction]}
            type="radio"
            value={2}
          />
          <label
            className="rating__label"
            htmlFor="star-2"
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            Rating 
            2
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-3"
            name="rating"
            onChange={[MockFunction]}
            type="radio"
            value={3}
          />
          <label
            className="rating__label"
            htmlFor="star-3"
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            Rating 
            3
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-4"
            name="rating"
            onChange={[MockFunction]}
            type="radio"
            value={4}
          />
          <label
            className="rating__label"
            htmlFor="star-4"
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            Rating 
            4
          </label>
          <input
            checked={false}
            className="rating__input"
            id="star-5"
            name="rating"
            onChange={[MockFunction]}
            type="radio"
            value={5}
          />
          <label
            className="rating__label"
            htmlFor="star-5"
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            Rating 
            5
          </label>
        </div>
      </div>
      <div
        className="add-review__text"
      >
        <textarea
          className="add-review__textarea"
          name="comment"
          onChange={[MockFunction]}
          placeholder="Review text"
          value="comment"
        />
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;
